### PR TITLE
Fixed #29206 -- Fixed PasswordResetConfirmView crash when the URL contains a non-UUID where one is expected.

### PR DESCRIPTION
**Pull request by MattBlack85**
https://code.djangoproject.com/ticket/29206

### CODE REVIEW

File: django/contrib/auth/views.py [hunk 1]
```diff
@@ -285,7 +286,8 @@ def get_user(self, uidb64):
             # urlsafe_base64_decode() decodes to bytestring
             uid = urlsafe_base64_decode(uidb64).decode()
             user = UserModel._default_manager.get(pk=uid)
-        except (TypeError, ValueError, OverflowError, UserModel.DoesNotExist):
+        except (TypeError, ValueError, OverflowError, UserModel.DoesNotExist,
+                ValidationError):
```
Comment (carltongibson) [inline django/contrib/auth/views.py]: Please put on a single line.

File: tests/auth_tests/models/custom_user.py [hunk 2]
```diff
@@ -74,6 +76,7 @@ class RemoveGroupsAndPermissions:
     fields from the AbstractUser class, so they don't clash with the
     related_name sets.
     """
+
```
Comment (carltongibson) [inline tests/auth_tests/models/custom_user.py]: Please remove this unrelated change.

File: tests/auth_tests/test_templates.py [hunk 3]
```diff
@@ -53,6 +54,23 @@ def test_PasswordResetConfirmView_valid_token(self):
         self.assertContains(response, '<title>Enter new password</title>')
         self.assertContains(response, '<h1>Enter new password</h1>')
 
+    @override_settings(AUTH_USER_MODEL='auth_tests.CustomUserWithUUIDPk')
+    def test_PasswordResetConfirmView_invalid_uid(self):
+        # https://code.djangoproject.com/ticket/29206
```
Comment (carltongibson) [inline tests/auth_tests/test_templates.py]: Please replace this with a docstring, such as: 

```
"""
ValidationErrors in user lookup are handled correctly (#29206).
"""
```

File: tests/auth_tests/test_templates.py [hunk 4]
```diff
@@ -53,6 +54,23 @@ def test_PasswordResetConfirmView_valid_token(self):
...
+        rf = RequestFactory()
+        user = CustomUserWithUUIDPk.objects.create_user('jsmith', 'jsmith@example.com', 'pass')
+        user = authenticate(username=user.username, password='pass')
+        request = rf.get('/somepath/')
+        request.user = user
+        client = PasswordResetConfirmClient()
+        default_token_generator = PasswordResetTokenGenerator()
+        token = default_token_generator.make_token(user)
+        uidb64 = urlsafe_base64_encode(str(user.pk).encode()).decode() + 'Mh17ag01'
```
Comment (carltongibson) [inline tests/auth_tests/test_templates.py]: I don't think you need to actually create the user (or customise the request) here. If this line reads: 

```
        uidb64 = urlsafe_base64_encode(str('NOT_A_VALID_UUID').encode()).decode()
```

... then the code reduces to: 

```
        client = PasswordResetConfirmClient()
        default_token_generator = PasswordResetTokenGenerator()
        token = default_token_generator.make_token(self.user)
        uidb64 = urlsafe_base64_encode(str('NOT_A_VALID_UUID').encode()).decode()
        url = reverse('password_reset_confirm', kwargs={'uidb64': uidb64, 'token': token})
        response = client.get(url)
        # ... continue with asserts... 
```

**Comment by MattBlack85** [comment]
@carltongibson thanks for your review, I applied the changes as requested

**Comment by carltongibson** [comment]
@MattBlack85 Wow. That was quick. Thanks.

File: tests/auth_tests/test_templates.py [hunk 5]
```diff
@@ -53,6 +53,19 @@ def test_PasswordResetConfirmView_valid_token(self):
...
         self.assertContains(response, '<h1>Enter new password</h1>')
 
+    def test_PasswordResetConfirmView_invalid_uid(self):
+        """
+        ValidationError in user lookups are handled correctly (#29206).
+        """
+        client = PasswordResetConfirmClient()
+        default_token_generator = PasswordResetTokenGenerator()
+        token = default_token_generator.make_token(self.user)
+        uidb64 = urlsafe_base64_encode('NOT_A_VALID_STRING'.encode()).decode()
```
Comment (carltongibson) [inline tests/auth_tests/test_templates.py]: Sorry. Could this say `"NOT_A_VALID_UUID"`? I think that would be clearer.

Comment (MattBlack85) [reply tests/auth_tests/test_templates.py]: I'm wondering, since it's not strictly related to uuids as you correctly pointed out, but to uid in general, if uid would be more appropriate here?

Comment (carltongibson) [reply tests/auth_tests/test_templates.py]: But it's only because it's not a valid UUID that the ValidationError is raised, which (that it's then caught) is what the test is checking. 

Hang on... you broke the test...

Comment (MattBlack85) [reply tests/auth_tests/test_templates.py]: Will re-add the custom user model tomorrow then👌

**Comment by timgraham** [comment]
Can't you use the existing `UUIDUser` model?

File: tests/auth_tests/test_templates.py [hunk 6]
```diff
@@ -53,6 +54,25 @@ def test_PasswordResetConfirmView_valid_token(self):
         self.assertContains(response, '<title>Enter new password</title>')
         self.assertContains(response, '<h1>Enter new password</h1>')
 
+    @override_settings(AUTH_USER_MODEL='auth_tests.UUIDUser')
+    def test_PasswordResetConfirmView_invalid_uuid(self):
+        """
+        ValidationError in user lookups are handled correctly (#29206).
+        """
+        rf = RequestFactory()
```
Comment (carltongibson) [inline tests/auth_tests/test_templates.py]: @MattBlack85 My previous point is that I don't think the steps between here and line 66 (`request.user = user`) are necessary. 

If you just begin with `client = ` we still get the `ValidationError` thrown when attempting to look up the user instance.

Comment (MattBlack85) [reply tests/auth_tests/test_templates.py]: you're perfectly right, I thought I still needed the entire machinery with the custom user

**Comment by timgraham** [comment]
I moved the tests from `test_templates.py` to `test_views.py` since that's what's being tested.

Carlton, since I see you suggested the test docstring, remember that ticket references should be reserved for obscure issues (from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style)). I think the behavior here is straigthtforward and doesn't need the ticket reference.

**Comment by carltongibson** [comment]
Good. Thanks for the pointer Tim. 👍🏽
